h2/src/main/org/h2/table/TableFilter.java

```diff
@@ -197,7 +198,7 @@ public PlanItem getBestPlanItem(Session s, TableFilter[] filters, int filter) {
             item1.setIndex(table.getScanIndex(s, null, filters, filter, sortOrder));
             item1.cost = item1.getIndex().getCost(s, null, filters, filter, sortOrder);
         }
-        int len = table.getColumns().length;
+        int len = table.getColumnIdCount();
```

Comment:
Both visible columns and system ones may appear on the same Row and their indexes must not clash. 
Current row size is only enough to hold visible columns, but not additional system ones.
This new getColumnIdCount extends the Row's size to ensure it does have space for both visible and invisible columns.